tfhe/src/integer/ciphertext/boolean_value.rs

```diff
@@ -0,0 +1,191 @@
+///
+/// ```
+/// use tfhe::integer::{gen_keys_radix, BooleanValue};
+/// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+///
+/// // We have 4 * 2 = 8 bits of message
+/// let size = 4;
+/// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2_KS_PBS, size);
+///
+/// let a = 128u8;
+/// let b = 55u8;
+///
+/// let ct_a = cks.encrypt(a);
+/// let ct_b = cks.encrypt(b);
+///
+/// let ct_res = sks.ge_parallelized(&ct_a, &ct_b);
+/// // Convert the boolean value to a RadixCiphertext of size blocks
+/// // so we can use it in operations
+/// let ct_res = ct_res.into_radix(size, &sks);
+/// // Decrypt:
+/// let dec: u8 = cks.decrypt(&ct_res);
+/// assert_eq!(u8::from(a >= b), dec);
+/// ```
+///
+/// ## Wrapping a ciphertext in a BooleanValue
+/// ```
+/// use tfhe::integer::{gen_keys_radix, BooleanValue};
+/// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+///
+/// // We have 4 * 2 = 8 bits of message
+/// let size = 4;
+/// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2_KS_PBS, size);
+///
+/// let clear_condition = 1u8;
+/// let a = 128u8;
+/// let b = 55u8;
+///
+/// let ct_a = cks.encrypt(a);
+/// let ct_b = cks.encrypt(b);
+/// let ct_condition = cks.encrypt(clear_condition);
+///
+/// // if_then_else requires a BooleanValue, so we wrap ct_condition
+/// // we wrap a reference to ct_condition to show that its possible to do so.
+/// let wrapped_condition = BooleanValue::new(&ct_condition);
+/// let ct_res = sks.if_then_else_parallelized(&wrapped_condition, &ct_a, &ct_b);
+///
+/// // Decrypt:
+/// let dec: u8 = cks.decrypt(&ct_res);
+/// assert_eq!(if clear_condition != 0 { a } else { b }, dec);
+/// assert_ne!(ct_a, ct_res);
+/// assert_ne!(ct_b, ct_res);
+/// ```
+#[derive(Clone, Deserialize, Serialize)]
+pub struct BooleanValue<Inner>(pub Inner);
+
+impl<Inner> BooleanValue<Inner> {
+    /// Simple constructor used to wrap a ciphertext or reference to a ciphertext
+    /// to mark it as encrypting 0 or 1
+    pub fn new(inner: Inner) -> Self {
+        Self(inner)
+    }
+}
+
+impl BooleanValue<Ciphertext> {
+    pub fn into_radix<T>(self, num_blocks: usize, sks: &ServerKey) -> T
+    where
+        T: IntegerRadixCiphertext,
+    {
+        let mut radix_ct = RadixCiphertext::from_blocks(vec![self.0]);
+        let missing_blocks = num_blocks.saturating_sub(1);
+        sks.extend_radix_with_trivial_zero_blocks_msb_assign(&mut radix_ct, missing_blocks);
+        T::from_blocks(radix_ct.blocks)
+    }
+}
+
+impl<Inner> BooleanValue<Inner>
+where
+    Inner: IntegerRadixCiphertext,
+{
+    pub fn into_radix(self, num_blocks: usize, sks: &ServerKey) -> Inner {
+        let mut radix_ct = RadixCiphertext::from_blocks(self.0.into_blocks());
+        let missing_blocks = num_blocks.saturating_sub(radix_ct.blocks().len());
+        sks.extend_radix_with_trivial_zero_blocks_msb_assign(&mut radix_ct, missing_blocks);
+        Inner::from_blocks(radix_ct.blocks)
+    }
```

Comment:
Does this handle the case where the boolean value comes from a u32 and then is casted to a u16 ? i.e. the blocks may need trimming ?

what about always storing a RadixCiphertext that always has a single block ? or always a shortint Ciphertext ?

I don't remember what we said on that when we discussed it recently (a single shortint block vs something else).

Reply:
You're right it does not handle the case where the target num blocks is smaller than the current number.

If we were to force the use of one representation I would be in favor of single block (and not a Vec with single block)

However the idea was for this wrapper to be as frictionless as possible, and so allow it to wrap a RadixCiphertext or a &RadixCiphertext in case that is was the user have its 'hands'

Reply:
ah right the ref case is a bit annoying, but I guess we could have BooleanBlock<&Ciphertext> and you could do something like from_ct_ref(&radix_ct) { return BooleanVec(&radix_ct.blocks()[0]) } ?

As always frictionless is good, just need to find a design that avoids as many mistakes as possible